src/lib.rs

```diff
@@ -89,6 +89,59 @@ impl<K: std::hash::Hash + PartialEq, V> HashTable<K, V> {
     }
 }
 
+pub struct HashTableIterator<'a, K, V> {
+    current_bucket: Option<&'a Vec<(K, V)>>,
+    elements_iterator: Box<dyn Iterator<Item = &'a (K, V)> + 'a>,
+    buckets_iterator: Box<dyn Iterator<Item = &'a Vec<(K, V)>> + 'a>,
+}
+
+impl<'a, K, V> IntoIterator for &'a HashTable<K, V> {
+    type Item = &'a (K, V);
+
+    type IntoIter = HashTableIterator<'a, K, V>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        let mut buckets_iterator = self.buckets.iter();
+        let current_bucket = buckets_iterator.next();
+        // first elements iterator needs to be initialized
+        let elements_iterator = match current_bucket {
+            Some(b) => b.iter(),
+            None => [].iter(),
+        };
+        HashTableIterator {
+            current_bucket,
+            elements_iterator: Box::new(elements_iterator),
```

Comment:
for both iterators, I understood that an indirection (reference, `Box`, etc.) was required for storing the `Iterator` traits in the struct since they are not `Sized` at compile time. Is there a way to remove the `Box`ing or is it necessary?

Reply:
See: https://stackoverflow.com/questions/47838596/how-do-i-have-a-structs-field-be-an-iterator-over-t-elements/47838741
Generics solution wasn't achievable in reasonable amount of time given my skill level so leaving as is.